File: app/src/main/java/com/example/foodorderingapp/activity/SignUpActivity.java
package com.example.foodorderingapp.activity;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.example.foodorderingapp.databinding.ActivitySignUpBinding;
import com.example.foodorderingapp.model.User;
import io.realm.Realm;
import java.util.UUID;

public class SignUpActivity extends AppCompatActivity {
    private Realm realm;
    private ActivitySignUpBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySignUpBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        Realm.init(this);
        realm = Realm.getDefaultInstance();

        binding.alreadyHaveButton.setOnClickListener(v -> {
            Intent intent = new Intent(SignUpActivity.this, LoginActivity.class);
            startActivity(intent);
        });

        binding.createAccount.setOnClickListener(v -> {
            String name = binding.nameRegistration.getText().toString();
            String email = binding.emailRegistration.getText().toString();
            String password = binding.passwordRegistration.getText().toString();

            registerUser(name, email, password);
        });
    }

    private void registerUser(String username, String email, String password) {
        if (username.isEmpty() || email.isEmpty() || password.isEmpty() || !isValidEmail(email)) {
            Toast.makeText(this, "Invalid username, email, or password", Toast.LENGTH_SHORT).show();
            return;
        }

        User existingUser = realm.where(User.class).equalTo("userEmail", email).findFirst();

        if (existingUser != null) {
            Toast.makeText(this, "Username already exists!", Toast.LENGTH_LONG).show();
        } else {
            realm.executeTransactionAsync(realm -> {
                User newUser = realm.createObject(User.class, UUID.randomUUID().toString());
                newUser.setUserName(username);
                newUser.setUserEmail(email);
                newUser.setUserPassword(password);
                newUser.setUserType("customer");

                Handler handler = new Handler(Looper.getMainLooper());
                handler.post(() -> {
                    Toast.makeText(SignUpActivity.this, "User registered successfully!", Toast.LENGTH_LONG).show();
                    binding.nameRegistration.setText("");
                    binding.emailRegistration.setText("");
                    binding.passwordRegistration.setText("");
                });

                startActivity(new Intent(SignUpActivity.this, LoginActivity.class));
                finish();
            });
        }
    }

    private boolean isValidEmail(String email) {
        return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close();
    }
}

File: app/src/main/java/com/example/foodorderingapp/SignUpActivity.kt
package com.example.foodorderingapp

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import com.example.foodorderingapp.databinding.ActivityLoginBinding
import com.example.foodorderingapp.databinding.ActivitySignUpBinding
import com.example.foodorderingapp.model.UserModel
import com.google.firebase.Firebase
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.auth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.database

class SignUpActivity : AppCompatActivity() {
    private lateinit var email: String
    private lateinit var username: String
    private lateinit var password: String
    private lateinit var auth: FirebaseAuth
    private lateinit var database: DatabaseReference

    private val binding: ActivitySignUpBinding by lazy {
        ActivitySignUpBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        //firebase authentication

        binding.userCreateAccount.setOnClickListener {
            auth = Firebase.auth
            database = Firebase.database.reference

            email = binding.userEmail.text.toString().trim()
            password = binding.userPassword.text.toString().trim()
            username = binding.userName.text.toString().trim()


            if(email.isBlank() || password.isBlank() || username.isBlank()) {
                Toast.makeText(this, "Please Fill All Details", Toast.LENGTH_SHORT).show()
            } else {
                createAccount(email, password)
            }
        }

        binding.alreadyHaveButton.setOnClickListener {
            val intent = Intent(this, LoginActivity::class.java)
            startActivity(intent)
        }
    }

    private fun createAccount(email: String, password: String) {
        auth.createUserWithEmailAndPassword(email, password).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                Toast.makeText(this, "Account Created Successfully", Toast.LENGTH_SHORT).show()
                saveUserData()
                val intent = Intent(this, LoginActivity::class.java)
                startActivity(intent)
                finish()
            } else {
                Toast.makeText(this, "Account Creation Failed", Toast.LENGTH_SHORT).show()
                Log.d("signUpfailure", "createAccountFailure", task.exception)
            }
        }
    }

    private fun saveUserData() {
        email = binding.userEmail.text.toString().trim()
        password = binding.userPassword.text.toString().trim()
        username = binding.userName.text.toString().trim()

        val user = UserModel(username, email, password)
        val userId = FirebaseAuth.getInstance().currentUser!!.uid
        database.child("user").child(userId).setValue(user)
    }
}